### ISSUE TITLE
Ambiguous threads dependency

### ISSUE THREAD

**Issue by LRN** [mesonbuild/meson#553]
`dependency('threads')` is poorly-documented.

Looking at the code (0.31.0), it seems that all that `dependency('threads')` does is create a dependency object that has `needs_threads() == True`.

`needs_threads() == True` makes the implementation add `compiler.thread_flags()` and `compiler.thread_link_flags()` to the flags that are actually used.

`compiler.thread_flags()` is `-pthread` for C and an empty string for C++
`compiler.thread_link_flags()` is `-pthread` for C and an empty string for C++

`<threading-rant>`

This does not seem like a good idea. A better approach is to focus on threading APIs.

I.e. `meson.build` would request threading API `X` for language `Y`. Meson will then do appropriate lookups and provide appropriate flags.

Threading APIs that i know of:

Name  | OS | library      | flags     | languages
------|----|--------------|-----------|----------
POSIX | GNU| libc         | -pthread? | C/C++, bindings
POSIX |*BSD| libpthread   | -lpthread | C/C++, bindings
POSIX |W32 | libpthread   | -lpthread | C/C++, bindings
W32   |W32 | libkernel32  | -lkernel32| C/C++, bindings
Python|Any | Python stdlib|           | Python

and so on for C#, Rust and other languages (Fortran?).

That said, it doesn't make much sense to check for these, as, for example, Python threading API is always part of Python stdlib, and W32 threading API is always available on W32 (-lkernel32 is hardcoded into GCC spec files, IIRC, you have to make an *effort* to *not* to link to it; dunno about MSVC).

POSIX threads is the only API that needs actual checking (as it might be in libc or in libpthread or somewhere else, in exotic OSes, and it might need special compiler flags).

Projects can also support multiple threading APIs (glib supports pthreads and W32 threads, at least) in their code. Which one is used is usually decided at configuration time by a configure option.

I would propose to document that `dependency('threads')` looks for POSIX threads (renaming it to `dependency('pthreads')` would probably break backward compatibility). Then actually either *do* check which flags are needed to link a pthread program (a simple `pthread_create()`) *or* hardcode that info into specific compiler profiles (i.e. `-lpthread` is sufficient for all MinGW flavors, as they traditionally come with some kind of pthreads implementation; standard GCC just needs `-pthread`). It should be overridable (in case someone wants to supply a threading library that implements POSIX threads, but does not provide a library named `libpthread` and does not provide a header named `/include/pthread.h`; it could be instead `libmythread` and `/include/mythread/pthread.h`, which means `-lmythread -I/include/mythread`).

Someone who knows obscure POSIX-ish platforms would need to verify the above. Digging into autotools source code (which does contain info about obscure OS quirks) is probably not an option, as it's GPL-ed, while meson is Apache2.0-ed.

As for the rest of the APIs, i don't know them well enough (even for W32 threading API, which i do know, there may be special MSVC flags that i'm not aware of, because i don't use MSVC) to judge whether it's enough to just imply their availability for a specific platform, or whether they need some kind of flags as well.

Looking back, i think this is an OS quirk - the only case where you *need* special compiler flags for threading support is when it comes from the OS. When it's in stdlib of a language, you don't need to do anything special, and stdlib uses appropriate OS API under the hood. When it's just a 3rd-party library, you treat it as such, and the library uses appropriate OS API under the hood. Therefore, the number of *really quirky* threading APIs would actually be roughly equal to the number of OSes supported by meson, which right now is, like, 4?

`</threading-rant>`

**Comment by jpakkane** [comment]
I'm in bit of a rush but quickly:

- threads adds -pthread even for C++ 
- the wrench in the works is MSVC, Windows does native threaded libraries bizarrely (I don't know the nitty gritty details, though)

**Comment by LRN** [comment]
> threads adds -pthread even for C++ 

I stand corrected. The compilers that has empty string for thread flags is actually MSVC C and C++ compilers, not the generic C++ compiler (which actually inherits '-pthread' from the generic C compiler).

> the wrench in the works is MSVC, Windows does native threaded libraries bizarrely (I don't know the nitty gritty details, though)

You've mentioned that before. What exactly is so bizarre about W32 threading? Maybe some keywords for me to google up?

**Comment by jpakkane** [comment]
I don't know the specifics but with Visual Studio you can compile your code with multithreading support or without it. AFAICT these two modes are ABI incompatible, that is, you can't mix code that uses both types in a single process. VS also has debug and nondebug modes which might or might not work together with this.

As an example, Meson's test case that builds a Python extension overrides the default build type because the default Win packages that Python ships are not debug enabled so an extension built in debug mode would not work with it (the test suite uses the host Python to run the built module).

**Comment by LRN** [comment]
[MSDN on CRTs](https://msdn.microsoft.com/en-us/library/1bh5ewb2.aspx) claims that threading is supported in all versions of CRT starting from MSVS2005, so there's no need to chose different CRT versions depending on whether the application is expected to be multithreaded or not. This might be what you are referring to.

Since you've mentioned Python, there's another unrelated wrinkle - official Python binaries are MSVS-built (using some kind of MSVCR10XX), but MSYS2 repositories contain Python that is built with MinGW (using MSVCRT by default, IIRC). Mixing different CRTs (which, as you've pointed out, also includes debug- and non-debug-versions of it for MSVS) is a never-ending source of gotchas.

**Comment by jpakkane** [comment]
I checked Boost 1.61 and the binaries they ship .lib files for VS2010 called `foo-s.lib`, `foo-mt.lib` and a few others. So something related to threads seems to be happening here. There seems to be only one dll, though.

**Comment by adrianjohnson** [comment]
I've also found dependency('threads') to be ambiguous. It doesn't tell me what threads implementation it is using or provide any way to prefer a particular API on platforms that support multiple threading APIs.

The CMake FindThreads tells me what thread API it is using and provides a way to prefer pthreads.

I checked the glib code to so how they handle it and after dependency('threads') they assume windows is using win32 threads and everything else is pthreads. That does not seem very reliable, particularly on mingw.

**Comment by adrianjohnson** [comment]
The problem with a generic 'threads' dependency is the application still needs to be coded to a specific API. Maybe it would be better to provide separate API specific thread dependencies "pthread', 'win32-threads' etc. Then the application can check for the APIs it supports in the order it prefers.

**Comment by tp-m** [comment]
Explicit API checks would work. And/or a dep getter to query which threads implementation/api was found. I need to know what API should be used in the source code.

**Comment by ePirat** [comment]
> The problem with a generic 'threads' dependency is the application still needs to be coded to a specific API. 

We have exactly the same problem for VLC and another project. The thread dependency is mostly useless the way it is as if a program is written for pthread, there is no way with it to ensure that it looks for pthread and not windows threads, so it breaks when cross-compiling for Windows, for example.
As a workaround we check if we are compiling for windows and then manually check with `dependency('pthread')` and `find_library('pthread')` but a better solution for this would be great.

**Comment by Ericson2314** [comment]
Making matters more complicated, I've started using https://github.com/lhmouse/mcfgthread/ with MinGW for C++ thread support, so it's not just posix or win32 there.

**Comment by ePirat** [comment]
@Ericson2314 well but that is clearly a "third party" library so using `find_library` seems reasonable for that instead of expecting `dependency('threads')` to somehow find it?

Of course if we decide to add a way for `dependency('threads')` to find specific threading "flavor" we could support those too…

**Comment by Ericson2314** [comment]
@ePirat Well there is a GCC patch with that so that library so it would be baked into libstdc++ and friends. This blurs the 3rd-party 1st-party distinction.

**Comment by ePirat** [comment]
If I understand correctly mcfgthread is for std::thread implementation so not any of the C thread stuff? Maybe we could have a `dependency('threads', api: 'pthread')` or such to cover the different threading APIs?

**Comment by Ericson2314** [comment]
Or even `dependency('pthreads')`, there are multiple interfaces and multiple implementations of each, but you are exactly right: we shouldn't conflate the interfaces.

**Comment by ePirat** [comment]
The issue with `dependency('pthreads')` is that than we would need a lot of custom dependencies for the other threading interfaces/libs too, no?

And what do you mean by multiple interfaces? pthread is standardized so each implementation of it should have the same APIs? (Some have non-standard extensions but then I believe if you want them you should check explicitly for those functions instead of a specific pthread implementation?)

**Comment by Ericson2314** [comment]
> pthread is standardized so each implementation of it should have the same APIs?

Yes, agreed

> (Some have non-standard extensions but then I believe if you want them you should check explicitly for those functions instead of a specific pthread implementation?)

Yes, agreed

> And what do you mean by multiple interfaces?

Well I mean there are at least pthreads, C++ threads and possible other interfaces with multiple implementations. Neither pthreads or C++ threads is a subset of the other.

**Comment by ePirat** [comment]
Yes, I agree with that. Thats why I proposed the `api` kwarg for dependency('thread') to specify that. (Although `api` might not be the best name for that).

**Comment by Ericson2314** [comment]
@ePirat Right I am just saying that `dependency('pthreads')` and `dependency('c++threads')` could also work equally well, as dependencies are about interfaces not implementations.

**Comment by ePirat** [comment]
Sure, I just thought a kwarg would be easier as we already have lots of custom dependencies and it would clobber the namespace of these less and avoid that we ever clash with anything that actually exists with the same name but is something else…

**Comment by jpakkane** [comment]
At least on Linux C++ threads and pthreads are the one and the same. Both of them are enabled with `-pthread`.

**Comment by ePirat** [comment]
@jpakkane But does that matter for the user? `std::thread` has a specific API and the user will only care if that works or not, not how you enabled that with the compiler?

**Comment by Ericson2314** [comment]
Yeah it's best when we give the user only the interface they asked for, but as a fallback option we can expose more interface than that. Exposing both with Linux is like allowing one to mix `native: true` and `native: false` deps with native builds.

**Comment by lygstate** [comment]
I think it's better add dependency('pthreads') and dependency('c11threads') and dependency('cxx11threads') for different threading library.
The dependency('threads') introduce confusings more than the issues it solved.

**Comment by jscott0** [comment]
I would also like to point out that according to POSIX, the correct way to get the compiler flags necessary (`-pthread` is not in POSIX) is to use `getconf` or `confstr()`. In fact, one can use `getconf` or `sysconf()` to query for whether POSIX threads is supported at all

**Comment by eli-schwartz** [comment]
Sorry, which getconf value are you suggesting contains `-pthread` on systems where the compiler uses that flag?

**Comment by jscott0** [comment]
@eli-schwartz 
> Sorry, which getconf value are you suggesting contains `-pthread` on systems where the compiler uses that flag?

You're looking for `POSIX_V7_THREADS_CFLAGS` and `POSIX_V7_THREADS_LDFLAGS`. Note that these don't appear to be currently supported on my glibc system, but I believe at least the `confstr()` parameters are supported by musl libc. [This is a known glibc omission](https://sourceware.org/bugzilla//show_bug.cgi?id=25003)
